internal/app/handlers.go

```diff
@@ -116,3 +114,69 @@ func (h *Handlers) handleNormal(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Location", originalURL)
 	w.WriteHeader(http.StatusTemporaryRedirect)
 }
+
+func (h *Handlers) handleJSONShorten(w http.ResponseWriter, r *http.Request) {
+	// Устанавливаем базовый URL для сервиса
+	h.service.SetBaseURL(r)
+
+	// Проверяем метод
+	if r.Method != http.MethodPost {
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Проверяем Content-Type
+	if contentType := r.Header.Get("Content-Type"); contentType != "application/json" {
+		http.Error(w, "Content-Type must be application/json", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	// Читаем и парсим тело запроса
+	var req shortenRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http.Error(w, "Bad request: invalid JSON", http.StatusBadRequest)
+		return
+	}
+
+	originalURL := strings.TrimSpace(req.URL)
+	if originalURL == "" {
+		http.Error(w, "URL cannot be empty", http.StatusBadRequest)
+		return
+	}
+
+	// Пытаемся получить короткий URL, если он уже существует
+	if shortURL, exists := h.service.GetShortURLIfExists(originalURL); exists {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusCreated)
+
+		resp := shortenResponse{Result: shortURL}
+		if err := json.NewEncoder(w).Encode(resp); err != nil {
+			log.Printf("Failed to encode response: %v", err)
+			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		}
+		return
+	}
+
+	// Создаем новый короткий URL
+	shortURL, err := h.service.ShortenURL(originalURL)
+	if err != nil {
+		if errors.Is(err, service.ErrInvalidURL) || errors.Is(err, service.ErrEmptyURL) {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		log.Printf("Failed to shorten URL '%s': %v", originalURL, err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	}
```

Comment:
Я предложил бы упростить.
```suggestion
// Пытаемся получить короткий URL, если он уже существует
	var (
		shortURL string
		err      error
	)
	shortURL, exists := h.service.GetShortURLIfExists(originalURL)
	if !exists {
		// Создаем новый короткий URL
		shortURL, err = h.service.ShortenURL(originalURL)
		if err != nil {
			if errors.Is(err, service.ErrInvalidURL) || errors.Is(err, service.ErrEmptyURL) {
				http.Error(w, err.Error(), http.StatusBadRequest)
				return
			}

			log.Printf("Failed to shorten URL '%s': %v", originalURL, err)
			http.Error(w, "Internal server error", http.StatusInternalServerError)
			return
		}
	}
```